okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -239,48 +239,44 @@ public Void visit(QosException.Throttle exception) {
 
             @Override
             public Void visit(QosException.RetryOther exception) {
-                if (maxNumRelocations <= 0) {
-                    callback.onFailure(call,
-                            new IOException("Exceeded the maximum number of allowed "
-                                    + "redirects for initial URL: " + call.request().url()));
-                } else {
-                    Optional<HttpUrl> redirectTo = urls.redirectTo(request().url(),
-                            exception.getRedirectTo().toString());
-                    if (!redirectTo.isPresent()) {
-                        callback.onFailure(call,
-                                new IOException("Failed to determine valid redirect URL "
-                                        + "for '" + exception.getRedirectTo() + "' and base URLs "
-                                        + urls.getBaseUrls()));
-                    } else {
-                        Request redirectedRequest = request().newBuilder()
-                                .url(redirectTo.get())
-                                .build();
-                        client.newCallWithMutableState(
-                                redirectedRequest, backoffStrategy, maxNumRelocations - 1)
-                                .enqueue(callback);
-                    }
-                }
-                return null;
+                return redirectToUrl(callback, call,
+                        // Redirect to the URL specified by the exception.
+                        () -> urls.redirectTo(request().url(), exception.getRedirectTo().toString()),
+                        () -> "Failed to determine valid redirect URL for '" + exception.getRedirectTo() + "' and base "
+                                + "URLs " + urls.getBaseUrls());
             }
 
             @Override
             public Void visit(QosException.Unavailable exception) {
-                Optional<Duration> backoff = backoffStrategy.nextBackoff();
-                if (!backoff.isPresent()) {
-                    log.debug("Max number of retries exceeded, failing call");
-                    callback.onFailure(call,
-                            new IOException("Failed to complete the request due to a "
-                                    + "server-side QoS condition: 503", exception));
-                } else {
-                    log.debug("Rescheduling call after backoff",
-                            SafeArg.of("backoffMillis", backoff.get().toMillis()));
-                    scheduleExecution(callback, backoff.get());
-                }
-                return null;
+                return redirectToUrl(callback, call,
```

Comment:
I'd like to retry another host *with backoff*

Reply:
Are you thinking of the single-node case here? I thought about this and ended up going this way just to index on the multi-node case - if we get a 503 from one node, we have no data on the health of the other nodes, thus we can safely assume (as we always do when we are issuing the first try of a request) that any other node is healthy right now, and we can try immediately with no backoff. But in the single-node case we'd definitely want to back off. But the logic of how many URLs there are shouldn't be present at this level... and I guess you could run into the case where _all_ nodes are unavailable due to load, and the client is getting bounced around, in which case it still makes sense to back off. And backing off the first time or two probably isn't the worst penalty to pay. Okay.